File: Labs/Lab7.py
import unittest

class Item:
    def __init__(self, weight, value):
        self.weight = weight
        self.value = value

def fractional_knapsack(items, capacity):
    items = sorted(items, key=lambda x: x.value / x.weight, reverse=True)
    total_value = 0

    for item in items:
        if capacity >= item.weight:
            total_value += item.value
            capacity -= item.weight
        else:
            total_value += item.value * (capacity / item.weight)
            break

    return total_value

# Unit Test
class TestFractionalKnapsack(unittest.TestCase):
    def test_example(self):
        items = [Item(10, 60), Item(20, 100), Item(30, 120)]
        self.assertAlmostEqual(fractional_knapsack(items, 50), 240.0, places=2)

    def test_no_capacity(self):
        items = [Item(10, 60), Item(20, 100)]
        self.assertAlmostEqual(fractional_knapsack(items, 0), 0.0)

if __name__ == "__main__":
    unittest.main()

File: Labs/Lab7.cpp
#include <vector>
#include <algorithm>
#include <iostream>
#include <cassert>

struct Item {
    int weight;
    int value;
};

bool compare(Item a, Item b) {
    double r1 = (double)a.value / a.weight;
    double r2 = (double)b.value / b.weight;
    return r1 > r2;
}

double fractionalKnapsack(std::vector<Item> items, int capacity) {
    std::sort(items.begin(), items.end(), compare);
    double totalValue = 0.0;

    for (const auto& item : items) {
        if (capacity >= item.weight) {
            totalValue += item.value;
            capacity -= item.weight;
        } else {
            totalValue += item.value * ((double)capacity / item.weight);
            break;
        }
    }

    return totalValue;
}

void testFractionalKnapsack() {
    std::vector<Item> items = {{10, 60}, {20, 100}, {30, 120}};
    assert(abs(fractionalKnapsack(items, 50) - 240.0) < 1e-6);
    std::cout << "All fractionalKnapsack tests passed!" << std::endl;
}

int main() {
    testFractionalKnapsack();
    return 0;
}
